core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
-                                }
-                            }, new ClusterStateObserver.ChangePredicate() {
-                                @Override
-                                public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
-                                                     ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
-                                    return newState.nodes().masterNodeId() != null;
-                                }
-
-                                @Override
-                                public boolean apply(ClusterChangedEvent event) {
-                                    return event.nodesDelta().masterNodeChanged();
-                                }
+                    retry(new MasterNotDiscoveredException(), masterNodeChangedPredicate());
+                } else {
+                    processBeforeDelegationToMaster(request, clusterState);
+                    transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
+                        @Override
+                        public Response newInstance() {
+                            return newResponse();
+                        }
+
+                        @Override
+                        public void handleResponse(Response response) {
+                            listener.onResponse(response);
+                        }
+
+                        @Override
+                        public String executor() {
+                            return ThreadPool.Names.SAME;
+                        }
+
+                        @Override
+                        public void handleException(final TransportException exp) {
+                            Throwable cause = exp.unwrapCause();
+                            if (cause instanceof ConnectTransportException) {
+                                // we want to retry here a bit to see if a new master is elected
+                                logger.debug("connection exception while trying to forward request to master node [{}], scheduling a retry. Error: [{}]",
+                                        nodes.masterNode(), exp.getDetailedMessage());
+                                retry(cause, masterNodeChangedPredicate());
+                            } else {
+                                listener.onFailure(exp);
                             }
-                    );
+                        }
+                    });
                 }
-                return;
             }
-            processBeforeDelegationToMaster(request, clusterState);
-            transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
+        }
+
+        private ClusterStateObserver.ChangePredicate retryableOrNoBlockPredicate() {
```

Comment:
retryableOrNoBlockPredicate depends on the current request being made. I chose a method so that the predicate object does not have to be created for every MasterNodeAction.